src/responder/kcm/kcmsrv_ccache_secdb.c

```diff
@@ -0,0 +1,1426 @@
+    ret = sss_sec_put(req, secret);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static errno_t sec_update_b64(TALLOC_CTX *mem_ctx,
+                              struct sss_sec_req *req,
+                              struct sss_iobuf *buf)
+{
+    errno_t ret;
+    TALLOC_CTX *tmp_ctx;
+    char *secret;
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    secret = sss_base64_encode(tmp_ctx,
+                               sss_iobuf_get_data(buf),
+                               sss_iobuf_get_size(buf));
+    if (secret == NULL) {
+        ret = EIO;
+        goto done;
+    }
+
+    ret = sss_sec_update(req, secret);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static const char *secdb_container_url_create(TALLOC_CTX *mem_ctx,
+                                              struct cli_creds *client)
+{
+    return talloc_asprintf(mem_ctx,
+                           KCM_SECDB_CCACHE_FMT,
+                           cli_creds_get_uid(client));
+}
+
+static const char *secdb_cc_url_create(TALLOC_CTX *mem_ctx,
+                                       struct cli_creds *client,
+                                       const char *secdb_key)
+{
+    return talloc_asprintf(mem_ctx,
+                           KCM_SECDB_CCACHE_FMT"%s",
+                           cli_creds_get_uid(client),
+                           secdb_key);
+}
+
+static const char *secdb_dfl_url_create(TALLOC_CTX *mem_ctx,
+                                        struct cli_creds *client)
+{
+    return talloc_asprintf(mem_ctx,
+                           KCM_SECDB_DFL_FMT,
+                           cli_creds_get_uid(client));
+}
+
+static errno_t kcm_ccache_to_secdb_kv(TALLOC_CTX *mem_ctx,
+                                      struct kcm_ccache *cc,
+                                      struct cli_creds *client,
+                                      const char **_url,
+                                      struct sss_iobuf **_payload)
+{
+    errno_t ret;
+    const char *url;
+    const char *key;
+    TALLOC_CTX *tmp_ctx;
+    struct sss_iobuf *payload;
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    key = sec_key_create(mem_ctx, cc->name, cc->uuid);
+    if (key == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    url = secdb_cc_url_create(tmp_ctx, client, key);
+    if (url == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = kcm_ccache_to_sec_input(mem_ctx, cc, client, &payload);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    ret = EOK;
+    *_url = talloc_steal(mem_ctx, url);
+    *_payload = talloc_steal(mem_ctx, payload);
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+struct ccdb_secdb {
+    struct sss_sec_ctx *sctx;
+};
+
+struct ccdb_secdb_state {
```

Comment:
I'm not sure if I understand how it's going to be used.

Reply:
Yes, sorry, it's confusing, so I added a comment. tl;dr we often don't need tevent state in the async requests, because they are only fake async, so we use this dummy structure.